### PR TITLE
5514 merge accounts

### PR DESCRIPTION
**Pull request by sekmiller**
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #5514 Add Admin API to merge accounts

## Pull Request Checklist

- [x] Unit [tests][x] none
- [x] Integration [tests][x]: Completed
- [x] Deployment requirements: None
- [x] [Documentation][docs] completed
- [x] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: http://guides.dataverse.org/en/latest/developers/sql-upgrade-scripts.html
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

**Comment by sekmiller** [comment]
Note that the MergeInAccountCommand does not have any permissions.  Is it enough to have the api endpoint under /admin?

File: doc/sphinx-guides/source/admin/user-administration.rst [hunk 1]
```diff
@@ -56,3 +56,14 @@ Using the API token 7ae33670-be21-491d-a244-008149856437 as an example:
...
+Combining Accounts
+------------------
+
+If a user has created multiple accounts and has been performed actions under both accounts that need to be preserved, these accounts can be combined.  One account can be merged into another account and all data associated with both accounts will be combined in the surviving account.
+
+This command::
+
+	curl -X PUT -d http://$server/api/admin/authenticatedUsers/$mergedIdentifier/mergeIntoUser/$continuingIdentifier
+
+moves account data from $mergedIdentifier into the account $continuingIdentifier and deletes the account of $mergedIdentifier.
```
Comment (pdurbin) [inline doc/sphinx-guides/source/admin/user-administration.rst]: I like how descriptive "continuing identifier" is but I would suggest that "to merge identifier" might make a little more sense that "merged identifier" to someone who is preparing to perform the merge by reading this guide.

File: src/main/java/edu/harvard/iq/dataverse/engine/command/impl/MergeInAccountCommand.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        List<RoleAssignment> consumedRAList = ctxt.roleAssignees().getAssignmentsFor(consumedAU.getIdentifier());
+        
+        for(RoleAssignment cra : consumedRAList) {
+            if(cra.getAssigneeIdentifier().charAt(0) == '@') {
+                //This still needs to check if the same Role Assignment exists in the baseRAList and if so... what?
+                boolean willDelete = false;
+                for(RoleAssignment bra : baseRAList) {
+                    //Matching on the id not the whole DVObject as I'm suspicious of dvobject equality
+                    if( bra.getDefinitionPoint().getId().equals(cra.getDefinitionPoint().getId())
+                        && bra.getRole().equals(cra.getRole())) { //TODO: Match on privateurltoken as well?? What does this do?
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/engine/command/impl/MergeInAccountCommand.java]: I'm happy to talk through the privateurltoken stuff if we'd like to remove this TODO from the code.

See also http://phoenix.dataverse.org/site/latest/site/apidocs/edu/harvard/iq/dataverse/privateurl/package-summary.html which starts with "The Private URL feature has been implemented as a specialized role assignment with an associated token that permits read-only access to the metadata and all files (regardless of if the files are restricted or not) of a draft version of a dataset."

And "A PrivateUrlUser is virtual in the sense that it does not have a row in the authenticateduser table. It exists so when a Private URL is enabled for a dataset, we can assign a read-only role ("member") to the identifier for the PrivateUrlUser." http://phoenix.dataverse.org/site/latest/site/apidocs/edu/harvard/iq/dataverse/authorization/users/PrivateUrlUser.html

In short, I don't think we need to worry about Private URLs for this pull request since they are not associated with the authenticateduser table. They're part of the dataset and shouldn't be touched with user accounts are merged.

**Comment by matthew-a-dunlap** [comment]
closing in favor of a branch combining #5514 and #5515 https://github.com/IQSS/dataverse/pull/5626
